java-checks/src/main/java/org/sonar/java/checks/OneExpectedCheckExceptionCheck.java

```diff
@@ -0,0 +1,180 @@
+  }
+
+  private void processAssertThrowsArguments(ExpressionTree expectedType, ExpressionTree executable) {
+    Optional<IdentifierTree> expectedException = getExpectedException(expectedType);
+    if (expectedException.isPresent() && executable.is(Tree.Kind.LAMBDA_EXPRESSION)) {
+      IdentifierTree expectedIdentifier = expectedException.get();
+      reportMultipleCallThrowingExceptionInTree(expectedIdentifier.symbolType(), ((LambdaExpressionTree) executable).body(), expectedIdentifier);
+    }
+  }
+
+  private static Optional<IdentifierTree> getExpectedException(ExpressionTree expectedType) {
+    if (expectedType.is(Tree.Kind.MEMBER_SELECT)) {
+      MemberSelectExpressionTree memberSelect = ((MemberSelectExpressionTree) expectedType);
+      ExpressionTree expression = memberSelect.expression();
+      if ("class".equals(memberSelect.identifier().name()) && expression.is(Tree.Kind.IDENTIFIER)) {
+        return Optional.of((IdentifierTree) expression);
+      }
+    }
+    return Optional.empty();
+  }
+
+  private static boolean isTryCatchIdiom(TryStatementTree tree) {
+    List<StatementTree> statementTrees = tree.block().body();
+    if (!statementTrees.isEmpty()) {
+      StatementTree lastElement = statementTrees.get(statementTrees.size() - 1);
+      if (lastElement.is(Tree.Kind.EXPRESSION_STATEMENT)) {
+        ExpressionTree expressionTree = ((ExpressionStatementTree) lastElement).expression();
+        if (expressionTree.is(Tree.Kind.METHOD_INVOCATION)) {
+          return JUNIT_FAIL_MATCHER.matches((MethodInvocationTree) expressionTree);
+        }
+      }
+    }
+    return false;
+  }
+
+  private void reportMultipleCallThrowingExceptionInTree(Type expectedException, Tree tree, Tree reportLocation) {
+    MethodInvocationThrowing visitor = new MethodInvocationThrowing(expectedException); //m.symbolType().isSubtypeOf(expectedType)
+    tree.accept(visitor);
+    List<MethodInvocationTree> methodInvocationTrees = visitor.methodInvocationTrees;
+    if (methodInvocationTrees.size() > 1) {
+      reportIssue(reportLocation,
+        "The tested checked exception can be raised from multiples call, it is unclear what is really tested.",
+        secondaryLocations(methodInvocationTrees),
+        null);
+    }
+  }
+
+  private static List<JavaFileScannerContext.Location> secondaryLocations(List<MethodInvocationTree> methodInvocationTrees) {
+    return methodInvocationTrees.stream()
+      .map(ExpressionUtils::methodName)
+      .map(expr -> new JavaFileScannerContext.Location("Method call", expr))
+      .collect(Collectors.toList());
+  }
+
+  private static class MethodInvocationThrowing extends BaseTreeVisitor {
+    List<MethodInvocationTree> methodInvocationTrees = new ArrayList<>();
+    private final Type expectedException;
+
+    MethodInvocationThrowing(Type expectedException) {
+      this.expectedException = expectedException;
+    }
+
+    @Override
+    public void visitMethodInvocation(MethodInvocationTree mit) {
```

Comment:
what about visiting newClasses as well? you can perfectly call constructors as well, which can define checkedExceptions:
```
new MyObjectThrowingException(methodThrowingException());
```